test/test_model.py

```diff
@@ -732,6 +734,133 @@ def assertBackendCalls(self, expected, *, reset=True):
         self.assertEqual(expected, self.harness._get_backend_calls(reset=reset))
 
 
+class TestContainers(unittest.TestCase):
+    def setUp(self):
+        meta = ops.charm.CharmMeta.from_yaml("""
+name: k8s-charm
+containers:
+  c1:
+    k: v
+  c2:
+    k: v
+""")
+        backend = ops.model._ModelBackend('myapp/0')
+        self.model = ops.model.Model(meta, backend)
+
+    def test_unit_containers(self):
+        containers = self.model.unit.containers
+        self.assertEqual(sorted(containers), ['c1', 'c2'])
+        self.assertEqual(len(containers), 2)
+        self.assertIn('c1', containers)
+        self.assertIn('c2', containers)
+        self.assertNotIn('c3', containers)
+        for name in ['c1', 'c2']:
+            container = containers[name]
+            self.assertIsInstance(container, ops.model.Container)
+            self.assertEqual(container.name, name)
+            self.assertIsInstance(container.pebble, ops.pebble.Client)
+        with self.assertRaises(KeyError):
+            containers['c3']
+
+        with self.assertRaises(RuntimeError):
+            other_unit = self.model.get_unit('other')
+            other_unit.containers
+
+    def test_unit_get_container(self):
+        unit = self.model.unit
+        for name in ['c1', 'c2']:
+            container = unit.get_container(name)
+            self.assertIsInstance(container, ops.model.Container)
+            self.assertEqual(container.name, name)
+            self.assertIsInstance(container.pebble, ops.pebble.Client)
+        with self.assertRaises(ops.model.ModelError):
+            unit.get_container('c3')
+
+        with self.assertRaises(RuntimeError):
+            other_unit = self.model.get_unit('other')
+            other_unit.get_container('foo')
+
+
+class TestContainerPebble(unittest.TestCase):
+    def setUp(self):
+        meta = ops.charm.CharmMeta.from_yaml("""
+name: k8s-charm
+containers:
+  c1:
+    k: v
+""")
+        backend = ops.model._ModelBackend('myapp/0')
```

Comment:
It feels a little weird to be mocking one dependency, but passing in the others. It might still be the correct way to do it, though.

Reply:
This seemed to me the simplest way to write this test, so keeping as is for now.